### ISSUE TITLE
www.dl-protecte.com

### ISSUE THREAD

**Issue by Bagu** [reek/anti-adblock-killer#3306]
URL : http://www.dl-protecte.com/123455601123455603123455610123455615123455617vciruwho9h
![dl-protecte](https://user-images.githubusercontent.com/41675/27281160-f175949a-54ea-11e7-82b7-e42a0c938f78.jpg)

Mozilla Firefox
uBlock Origin

Base filters + french easylist + malwares lists
This happen after 15-20 seconds

**Comment by uBlock-user** [comment]
Filter - `@@||dl-protecte.com^$elemhide`

@jspenguin2017

**Comment by Bagu** [comment]
Don't work, still have the screen ;)

**Comment by Helix255** [comment]
Same problem, maybe reek anti adblock is not updating ?

I can block this warning with greasemonkey and this script : https://greasyfork.org/en/scripts?utf8=%E2%9C%93&q=anti+adblock+reek

**Comment by uBlock-user** [comment]
Works fine. Clear the cache and site data and then refresh the page.

**Comment by jspenguin2017** [comment]
@uBlock-user Race condition... this should be fixed by generic solutions once the Userscript is refactored into extension...

**Comment by jspenguin2017** [comment]
@uBlock-user When you clear the cache, it removed the race condition and it gets killed by Script: 
![image](https://user-images.githubusercontent.com/7283682/27290265-ebb31238-54c9-11e7-9e18-45b893834c78.png)
But if you close DevTools and refresh, it comes back... The rule doesn't work. I'll assume this to be fixed for now because the extension shouldn't take too long... hopefully... 
I'm not sure what is going to happen on the AAK-Cont side, I guess I should pause the pipeline before I push out the change, and we'll figure something out later.

**Comment by uBlock-user** [comment]
Good find. Will we be able to still add solutions to the script via the interface ?

**Comment by jspenguin2017** [comment]
@uBlock-user No, that's the whole problem, having dynamic script introduce race condition as I must fetch the code from the background storage. You will be able to add solution to the extension directly by loading unpacked extension. It can be directly in a Git repository so you can rollback your changes and pull mine.

**Comment by uBlock-user** [comment]
That's bad news. 

> by loading unpacked extension.

That requires you to manually download the extension and manually load it in Chromium and not from 
 the WebStore, which is unacceptable to me as I load all my extensions from WebStore and nowhere else.

**Comment by jspenguin2017** [comment]
You can review the source code before loading it, it's far safer than loading from the store as you don't have a way to verify that the packed extension matches the opened source code. Google does scan the extension, but they don't enforce the extension to be exactly the same as its repository. With unpacked extension, you also have the control over what the update is, it only updates when you run `git pull` and you can view the difference to know exactly what changed. 

It's ultimately your choice, but I want to fix the race condition and this is the best way I found. 

I can force it to be synchronous if you don't care about CPU usage. Actually, Tampermonkey can already do that. But I do care about battery and CPU usage, so that's no go.

**Comment by uBlock-user** [comment]
I trust you, source code is not the issue here. The issue is that I have to download every time you push an update and then install it again and I'm not willing to do that.

> Tampermonkey can already do that. But I do care about battery and CPU usage, so that's no go.

I have personally tested TM Beta with that feature and CPU usage is only spike for a second and goes back to normal. I don't know about you, but it's acceptable to me. I would rather have a second spike than to do the manual labour for an extension.

**Comment by jspenguin2017** [comment]
@uBlock-user If you clone the repository, then all you need to do is run: 
```
git fetch --all
git reset --hard origin/master
```
And it will update the unpacked extension. You can schedule it in `crontab` so it updates automatically. Note that you may need to explicitly set the directory if the `current working directory` is not at the right place.

**Comment by jspenguin2017** [comment]
This is working consistently with my current prototype, no race condition at all 😄 
![image](https://user-images.githubusercontent.com/7283682/27315834-43d9b8c6-5539-11e7-9094-19bb3df2328a.png)

**Comment by uBlock-user** [comment]
@jspenguin2017 I don't have `git client` installed neither I will install another piece of software just for this. Can I still extract the script from the extension and install it in VM ?

**Comment by jspenguin2017** [comment]
@uBlock-user No, it will be highly integrated with the rest of the extension, I will stay in the secured content script context as much as possible to prevent conflicts with uBlock Origin. (Breaking `open` stops the Userscript from loading, that's exactly what I'm trying to avoid)

Also, script tags injected from content script bypasses Content Security Policy, which an Userscript cannot do.

**Comment by uBlock-user** [comment]
But will I still be able to do what I did with uBlock Extra ?

**Comment by jspenguin2017** [comment]
I might put uBlock Extra inside my extension, it's not decided yet.

**Comment by uBlock-user** [comment]
Okay nevermind, but will you upload the extension to WebStore for every update ?

**Comment by jspenguin2017** [comment]
@uBlock-user Maybe not every single one, once a day or two.

**Comment by Helix255** [comment]
great job ! work fine now! thks

**Comment by Bagu** [comment]
Does uBlock Extra exist for Firefox users ?

**Comment by uBlock-user** [comment]
uBlock Extra is not needed for FF and not compatible either, only for Chromium.

**Comment by Bagu** [comment]
Since a day, @@||dl-protecte.com^$elemhide don't work anymore...

Do you have an idea ?

**Comment by jspenguin2017** [comment]
Seems to work for me... 
![image](https://user-images.githubusercontent.com/7283682/28393266-1700f360-6ca3-11e7-980d-fdc9c40c769d.png)

**Comment by Bagu** [comment]
On some pages, no, it don't work. I install greasemonkey + reek script and it work again...I don't undestand.

**Comment by jspenguin2017** [comment]
Greasemonkey doesn't have race condition, but it will reach end of life by November. 
This should be handled by generic Antiblock.org defuser.
